vms/platformvm/validators/fee/fee.go

```diff
@@ -0,0 +1,161 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package fee
+
+import (
+	"math"
+
+	"github.com/ava-labs/avalanchego/vms/components/gas"
+
+	safemath "github.com/ava-labs/avalanchego/utils/math"
+)
+
+// CalculateExcess updates the excess value based on the target and current
+// usage over the provided duration.
+func CalculateExcess(
+	target gas.Gas,
+	current gas.Gas,
+	excess gas.Gas,
+	duration uint64,
+) gas.Gas {
+	if current < target {
+		return excess.SubPerSecond(target-current, duration)
+	}
+	return excess.AddPerSecond(current-target, duration)
+}
+
+// CalculateCost calculates how much to charge based on the dynamic fee
+// mechanism for [duration].
+func CalculateCost(
+	target gas.Gas,
+	current gas.Gas,
+	minPrice gas.Price,
+	excessConversionConstant gas.Gas,
+	excess gas.Gas,
+	duration uint64,
+) uint64 {
+	// If the current and target are the same, the price is constant.
+	if current == target {
+		price := gas.CalculatePrice(minPrice, excess, excessConversionConstant)
+		cost, err := safemath.Mul(duration, uint64(price))
+		if err != nil {
+			return math.MaxUint64
+		}
+		return cost
+	}
+
+	var (
+		cost uint64
+		err  error
+	)
+	for i := uint64(0); i < duration; i++ {
+		excess = CalculateExcess(target, current, excess, 1)
```

Comment:
(optional) Following on from comment above, this could be `feeCtx.UpdateExcess(1)`. On first pass I missed the lack of `:=` so didn't appreciate that it was an update.

Reply:
I didn't want `AdvanceTime` to mutate the `State` because `State` is going to be passed around in the execution and we don't want to run into a situation where we accidentally mutate it.